client/mysqltest.cc

```diff
@@ -3809,45 +3809,306 @@ void do_remove_files_wildcard(struct st_command *command)
   DBUG_VOID_RETURN;
 }
 
+/*
+  SYNOPSIS
+  recursive_copy
+  ds_source      - pointer to dynamic string containing source
+                   directory information
+  ds_destination - pointer to dynamic string containing destination
+                   directory information
+  DESCRIPTION
+  Recursive copy of <ds_source> to <ds_destination>
+*/
+
+static int recursive_copy(DYNAMIC_STRING *ds_source,
+                          DYNAMIC_STRING *ds_destination) {
+  /* Note that my_dir sorts the list if not given any flags */
+  MY_DIR *src_dir_info =
+      my_dir(ds_source->str, MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+  int error = 0;
+
+  /* Source directory exists */
+  if (src_dir_info) {
+    /* Note that my_dir sorts the list if not given any flags */
+    MY_DIR *dest_dir_info =
+        my_dir(ds_destination->str, MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+    /* Create destination directory if it doesn't exist */
+    if (!dest_dir_info) {
+      error = my_mkdir(ds_destination->str, 0777, MYF(0)) != 0;
+      if (error) {
+        my_dirend(dest_dir_info);
+        goto end;
+      }
+    } else {
+      /* Extracting the source directory name */
+      if (ds_source->str[strlen(ds_source->str) - 1] == '/') {
+        strmake(ds_source->str, ds_source->str,
+                strlen(ds_source->str) - 1);
```

Comment:
I personally think `strmake()` complicates things here. All you are effectively doing is:

```c
ds_source->str[strlen(ds_source->str)]= '\0';
ds_source->length--;
```

Nothing really wrong with the way you are doing it, mostly just a comment on possibly simplifying it.